lm_eval/tasks/squad.py

```diff
@@ -64,23 +63,65 @@ def process_results(self, doc, results):
         :param results:
             The results of the requests created in construct_requests.
         """
-        # TODO: implement evaluation.
-        raise NotImplementedError('Evaluation not implemented')
+        squad_metric = datasets.load_metric("squad_v2")
+
+        continuation, is_unanswerable = results
+
+        logprob_unanswerable, is_greedy = is_unanswerable
+
+        no_answer_probability = exp(logprob_unanswerable)
+        
+        predictions = [{
+            'id': doc['id'],
+            'prediction_text': continuation,
+            'no_answer_probability': no_answer_probability,
+        }]
+
+        references = [{
+            'id': doc['id'],
+            'answers': doc['answers'],
+        }]
+
+        metrics = squad_metric.compute(predictions=predictions, references=references)
+
+        metrics.pop('total', None)
+        metrics.pop('HasAns_total', None)
+        metrics.pop('NoAns_total', None)
+        metrics.pop('best_exact_thresh', None)
+        metrics.pop('best_f1_thresh', None)
+
+        return metrics
 
     def aggregation(self):
         """
         :returns: {str: [float] -> float}
             A dictionary where keys are the names of submetrics and values are 
             functions that aggregate a list of metrics
         """
-        # TODO: implement evaluation.
-        raise NotImplementedError('Evaluation not implemented')
+        return { 
+            'exact': mean, # Exact match (the normalized answer exactly match the gold answer)
+            'f1': mean, #  The F-score of predicted tokens versus the gold answer
+            'HasAns_exact': mean, # Exact match (the normalized answer exactly match the gold answer)
+            'HasAns_f1': mean, # The F-score of predicted tokens versus the gold answer
+            'NoAns_exact': mean, # Exact match (the normalized answer exactly match the gold answer)
+            'NoAns_f1': mean, # The F-score of predicted tokens versus the gold answer
+            'best_exact': mean, # Best exact match (with varying threshold)
+            'best_f1': mean, # Best F1 (with varying threshold)
+        }

```

Comment:
Are you 100% sure that calling `squad_metric.compute` and then taking the mean of each of these values is the same as passing the arrays of all the values into `squad_metric.compute` in the first place? I'm especially concerned about the F1 ones

Reply:
I'm not certain if there's a canonically correct way to aggregate F1 scores. How would you want to do this otherwise, given the `process_results` and `aggregation` APIs?

Reply:
Here is an example:
https://github.com/EleutherAI/lm-evaluation-harness/blob/7ad6bf456a8542d9cd6aa2a1eaf4c97acf7f88a7/lm_eval/tasks/glue.py#L362
https://github.com/EleutherAI/lm-evaluation-harness/blob/7ad6bf456a8542d9cd6aa2a1eaf4c97acf7f88a7/lm_eval/metrics.py#L24

You probably shouldn't use the `f1_score` function that's used there here, but instead you should define your own function.  Assuming the cost of repeat-calculating `squad_metric.compute` is basically negligible, you can have some `custom_aggregation_fn(golds_reals, attr)` and in `aggregation` just return `partial(custom_aggregation_fn, attr="NoAns_f1")` or something like that.